### PR TITLE
UI Tests for username change on /settings page.

### LINKED ISSUE #1168: E2E test - /settings page - Multiple issues available
## Context

We need to ensure that our core application features and user flows are fully covered by end-to-end (E2E) tests. We are using Playwright for these tests, and the workflow framework for E2E testing is already in place.

If you navigate to the e2e folder in the root of the project, you'll find that @JohnAllenTech has written numerous tests for the article page. These tests have been grouped into two categories: authenticated and unauthenticated. Please ensure that this pattern is followed when adding new tests.

## Expected Behavior

Below is a list of possible tests for the `/settings` page. These are suggested tests generated with the help of ChatGPT, meant to provide a ***starting point***. You should use your discretion to refine these tests and account for relevant edge cases. Ensure that tests for each input field align with the current Zod validation schema. 

Additionally, feel free to propose other valid tests that might apply specifically to the functionality of the settings page.

Tests should be standalone and focus on a single purpose, ensuring each test verifies only one specific behavior or functionality.

Test should be added to e2e/settings.spec.ts

Please leave a comment specifying the tests you plan to write, so others don't duplicate the same efforts. 
All PRs are accepted for Hacktoberfest.

**Note:** For this page, the unauthenticated test is only necessary to verify that the user is correctly redirected to the `/get-started` page.

**Note:** As multiple people may contribute to this issue, please ensure you pull the latest changes from the upstream and resolve any conflicts. [Sync the upstream](https://www.codu.co/articles/effortless-upstream-syncing-a-custom-alias-for-git-users-fowfliwo)

## Tests

1. Profile Picture Upload
Test: Upload a valid profile picture (JPG, PNG, or GIF).
Scenario: Simulate selecting an image file, uploading it, and verifying that it is displayed correctly in the UI.
Edge Case: Test for an invalid file format (e.g., PDF, BMP) and ensure proper error messaging.
2. Full Name Update
Test: Update the full name field and save the changes.
Scenario: Input a new name, click "Save Changes", and verify that the new name is updated and persisted across sessions.
Edge Case: Test for maximum length or invalid characters, if there are any validation rules.
3. Username Update
Test: Update the username field and save the changes.
Scenario: Input a new username, click "Save Changes", and verify that the username is reflected across the platform.
Edge Case: Ensure that no two users can have the same username and that errors are displayed for duplicates.
4. Bio Field Update
Test: Update the bio field with valid text.
Scenario: Input a bio within the character limit, save, and verify that it updates correctly.
Edge Case: Try entering more than 200 characters and verify that an error is displayed or that it prevents submission.
5. Location Update
Test: Change the location and save the changes.
Scenario: Input a new location (e.g., text or emoji), save, and verify the location is updated.
6. Website URL Update
Test: Enter a valid website URL.
Scenario: Add a valid URL, save, and ensure the link displays properly.
Edge Case: Test for invalid URLs (e.g., no protocol) and verify error handling.
7. Email Verification
Test: Verify email change process.
Scenario: Enter a new email in the "Update Email" field, click "Send Verification Email", and ensure that a verification email is sent.
Edge Case: Test for invalid email formats and ensure the system rejects them with appropriate error messages.
8. Notification Preferences
Test: Toggle notification settings (Allow notifications, Weekly Newsletter).
Scenario: Toggle these switches, save changes, and verify that the settings persist after reload or login.
Edge Case: Ensure settings are correctly toggled and do not change without user input.
9. Form Reset
Test: Reset the form.
Scenario: Change several fields, click "Reset", and verify that all fields revert to their original state.
10. Save Changes
Test: Save changes to all fields.
Scenario: Modify multiple fields (profile picture, name, username, etc.), click "Save Changes", and verify that all updates are correctly applied and saved.
11. Validation & Error Handling
Test: Input invalid data (e.g., empty required fields, incorrect email format).
Scenario: Attempt to save with invalid input and verify appropriate validation messages are displayed.
12. Accessibility
Test: Verify the page's accessibility (e.g., via keyboard navigation and screen readers).
Scenario: Ensure that all fields can be navigated with the keyboard and that screen readers can access field labels correctly.

## Screenshots

Settings page

### PR DESCRIPTION
**Pull request by petercr**
# ✨ Codu Pull Request 💻

Fixes #1168 
** Only Part 3.**

## Pull Request details

Added UI tests for `/settings` to:
- tests for input field & types
- tries to submit a bad user name
- checks for error visibility
- resets the form
- changes user name, saves, and checks for update

## Any Breaking changes
None

## Associated Screenshots
None, but the tests pass 😅 I promise!

## [Optional] What gif best describes this PR or how it makes you feel

Like this, only instead of Monday, Wed, Friday it's like start, middle, end 😆 
A dog that starts out grumpy, then get glad, then gets happy.

### CODE REVIEW

File: e2e/settings.spec.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import test from "@playwright/test";
+import { test, expect } from "@playwright/test";
 import { loggedInAsUserOne } from "./utils";
 
 test.describe("Unauthenticated setttings Page", () => {
```
Comment (JohnAllenTech) [inline e2e/settings.spec.ts:4]: Can we change this to authenticated settings page. Because the user is authenticated for these tests

File: e2e/settings.spec.ts [hunk 2]
```diff
@@ -10,6 +10,32 @@ test.describe("Authenticated settings Page", () => {
...
+    const errorMessage = page.locator('p:text-is("Username can only contain alphanumerics and dashes.")')
+    await expect(errorMessage).toBeVisible();
+    await expect(errorMessage).toHaveText('Username can only contain alphanumerics and dashes.');
+    // Reset the form 
+    await page.locator('button:has-text("Reset")').click();
+  
+    // Test that the input field can be filled with a valid value and saves it
+    await inputField.fill('codu-rules');
+    await page.locator('button[type="submit"]').click(); 
+    await expect(inputField).toHaveValue('codu-rules');
```
Comment (JohnAllenTech) [inline e2e/settings.spec.ts]: Will this wait for the username change to take effect? 

If you fill in the new username and then click submit.

Will this test see the new username you have entered as the value of the input and complete straight away?

Comment (petercr) [reply e2e/settings.spec.ts]: I mean the tests pass for it. But it could be just catching the still filled in form from before the reset. 

Would you like me to change it to a page refresh instead? 

Or would you like some kind of timeOut in-between the state change and checking for the change to persist?

I would be happy to add either one 👍🏻

File: e2e/settings.spec.ts (hunk 1 above)
Comment (petercr) [reply e2e/settings.spec.ts:4]: I'm not sure exactly what you're saying here? 

Did you want a separate file for auth vs. un-auth tests, because doesn't a user have to be logged in to edit their profile?

I guess I could use more detail on this 👌🏻

File: e2e/settings.spec.ts (hunk 2 above)
Comment (JohnAllenTech) [reply e2e/settings.spec.ts]: Yeah I am not sure is there a loading indicator for these types of change? It would be nice to see

1. Click submit
2. Loading indicator appears
3. Loading indicator hidden
4. Check new username has taken effect

File: e2e/settings.spec.ts (hunk 1 above)
Comment (JohnAllenTech) [reply e2e/settings.spec.ts:4]: Your new test is ran inside a test suite called 'Unauthenticated setttings Page'

But this is an authenticated test because users need to login to change their username.

If you just change the name of the test it will be fixed

**Review by JohnAllenTech — CHANGES_REQUESTED** [review]
Just made some general comments. I am not a maintainer here so everything is just a suggestion

**Comment by NiallJoeMaher** [comment]
@JohnAllenTech let me know when this is good to go since you are the overlord of tests.

**Comment by petercr** [comment]
@JohnAllenTech I made some changes and added the extra tests you had suggested. 
Tests now cover:

- waiting for toast to show
- waiting for toast to disappear
- full page refresh
- checking new value persists in input field

I did have to change a field in `/e2e/setup/setup.ts` to add `https://` to it.
Otherwise you get an error on the form in E2E testing.

I also added new class names to both the toast messages for UI testing.

Thanks

**Comment by petercr** [comment]
@JohnAllenTech I made some changes and added the extra tests you had suggested. 
Tests now cover:

- waiting for toast to show
- waiting for toast to disappear
- full page refresh
- checking new value persists in input field

I did have to change a field in `/e2e/setup/setup.ts` to add `https://` to it.
Otherwise you get an error on the form.

**Comment by petercr** [comment]
I made some changes and added the extra tests you had suggested. 
Tests now cover:

- waiting for toast to show
- waiting for toast to disappear
- full page refresh
- checking new value persists in input field

I did have to change a field in `/e2e/setup/setup.ts` to add `https://` to it.
Otherwise you get a error on the form.

File: e2e/settings.spec.ts [hunk 3]
```diff
@@ -10,6 +10,38 @@ test.describe("Authenticated settings Page", () => {
...
+    await expect(errorMessage).toHaveText('Username can only contain alphanumerics and dashes.');
+    // Reset the form 
+    await page.locator('button:has-text("Reset")').click();
+  
+    // Test that the input field can be filled with a valid value and saves it
+    await inputField.fill('codu-rules');
+    await page.locator('button[type="submit"]').click(); 
+    const toastError = page.locator('.toast-success');
+    await expect(toastError).toBeVisible();
+    await expect(toastError).toBeHidden();
```
[the remark this answers is not in the document]
Comment (petercr) [reply e2e/settings.spec.ts]: I am not sure there is even a loading spinner? 🤷🏻‍♂️

I can add the tests for user name length and the corresponding errors if you want me to.

I guess I will just add them anyways 👌🏻

[the remark this answers is not in the document]
Comment (petercr) [reply e2e/settings.spec.ts]: I have added the extra tests for min and max length on the user name input.
This should resolve this issue.

**Comment by JohnAllenTech** [comment]
> @JohnAllenTech I made some changes and added the extra tests you had suggested. 
> 
> Tests now cover:
> 
> 
> 
> - waiting for toast to show
> 
> - waiting for toast to disappear
> 
> - full page refresh
> 
> - checking new value persists in input field
> 
> 
> 
> I did have to change a field in `/e2e/setup/setup.ts` to add `https://` to it.
> 
> Otherwise you get an error on the form in E2E testing.
> 
> 
> 
> I also added new class names to both the toast messages for UI testing.
> 
> 
> 
> Thanks 

Hey @petercr apologies I was AFK for a few days. I'll take a look at this in a while and reply properly

File: e2e/settings.spec.ts [hunk 4]
```diff
@@ -10,6 +10,32 @@
...
+    const errorMessage = page.locator('p:text-is("Username can only contain alphanumerics and dashes.")')
+    await expect(errorMessage).toBeVisible();
+    await expect(errorMessage).toHaveText('Username can only contain alphanumerics and dashes.');
+    // Reset the form 
+    await page.locator('button:has-text("Reset")').click();
+  
+    // Test that the input field can be filled with a valid value and saves it
+    await inputField.fill('codu-rules');
+    await page.locator('button[type="submit"]').click(); 
+    await expect(inputField).toHaveValue('codu-rules');
```
[the remark this answers is not in the document]
Comment (CarolinaCobo) [reply e2e/settings.spec.ts]: @petercr this seems like a good addition as well if you have the time :)

**Comment by NiallJoeMaher** [comment]
Tests failing here.

**Comment by petercr** [comment]
Well crap 😕 that's no good. 
I will look into this over the weekend (next 2 days) and see if I can figure it out. 👌🏻

I wonder if this has to do with the newsletter sub function?
Either way I'm on it 👍🏻

**Comment by petercr** [comment]
When I run the test locally, they all pass 🤷🏼‍♂️ see pic below. 
I'm not sure but this may have to do with `/server/lib/newsletter.ts`.

I added extra code on line 39 to keep from getting the email API errors for the newsletter. 
You can see some controller errors from `posts.sidebarData` and  the tRPC error for `profile.edit`.

Here's what I added locally for this:

```js
 if (process.env.ENV === "local" || "development") {
    return true;
  }
```

But it might be simpler to go with:

```js
 if (process.env.ENV !== "production") {
    return true;
  }
```

Here is the image of the tests passing.

msrdc_ttyjHMMI1j
